### DISCUSSION TITLE
Integrate Dart Macros for Improved Developer Experience

### DISCUSSION THREAD

**Discussion by ghost** [vitusortner/floor#848] [Ideas]
This discussion proposes exploring the integration of Dart macros into the Floor library. Dart macros offer the potential to significantly enhance the developer experience by:

- Reducing boilerplate code: Macros can automate repetitive tasks, such as defining database entities and DAOs, leading to cleaner and more concise code.
- Improving code readability: By abstracting away complex logic, macros can make the codebase easier to understand and maintain.
- Enhancing compile-time safety: Macros can perform compile-time checks to catch errors early on, reducing the risk of runtime issues.

This proposal aims to initiate a discussion on the feasibility and potential benefits of incorporating Dart macros into Floor.  Specific areas of exploration could include:

- Automatic generation of database entities and DAOs from Dart classes.
- Simplified definition of database schemas.
- Compile-time validation of database queries.

I believe that leveraging Dart macros can significantly improve Floor's usability and efficiency, making it an even more attractive option for Flutter developers. I welcome any feedback and suggestions on this proposal.

**Possible Example:**

Instead of this:

```
@Entity
class Person {
  @PrimaryKey(autoGenerate: true)
  final int id;

  final String name;

  Person(this.id, this.name);
}
```

We could have something like this:

```
@DatabaseEntity
class Person {
  int id;
  String name;

  Person(this.id, this.name);
}
```

Where @DatabaseEntity would be a macro responsible for generating all the necessary boilerplate code for Floor.

**Resources:**

Dart Macros Documentation https://dart.dev/language/macros#how-macros-work
I'm eager to hear your thoughts on this! 😄

**Comment by ruicraveiro** [comment]
Sadly I think this no longer applies. https://medium.com/dartlang/an-update-on-dart-macros-data-serialization-06d3037d4f12
